### DISCUSSION TITLE
CMAL in config.yml

### DISCUSSION THREAD

**Discussion by maby200** [neuralhydrology/neuralhydrology#88] [Q&A] [answered]
In [Uncertainty estimation with DL for rainfall-runoff models](https://arxiv.org/pdf/2012.14295.pdf) 4 DL baselines where shown. 

I am using CMAL for estimating uncertainty and for that, the lines in `config.yml` for setting CMAL head are the following:

```yml
# prediction head [regression]. Define the head specific parameters below
head: cmal 

# ----> CMAL settings <----
n_distributions: 2
n_samples: 7500
negative_sample_handling: clip
mc_dropout: False
```

I have read the docs for NeauralHydrology and found those were the only specifications needed for running the model and getting uncertainty estimations. But in python, uncaught exception appears:

 

>   2022-06-11 00:38:11,194: Uncaught exception
>    Traceback (most recent call last):
>      File "/usr/local/bin/nh-run", line 8, in <module>
>        sys.exit(_main())
>      File "/usr/local/lib/python3.7/dist-packages/neuralhydrology/nh_run.py", line 43, in _main
>        start_run(config_file=Path(args["config_file"]), gpu=args["gpu"])
>      File "/usr/local/lib/python3.7/dist-packages/neuralhydrology/nh_run.py", line 76, in start_run
>        start_training(config)
>      File "/usr/local/lib/python3.7/dist-packages/neuralhydrology/training/train.py", line 23, in start_training
>        trainer.train_and_validate()
>      File "/usr/local/lib/python3.7/dist-packages/neuralhydrology/training/basetrainer.py", line 205, in train_and_validate
>        self._train_epoch(epoch=epoch)
>      File "/usr/local/lib/python3.7/dist-packages/neuralhydrology/training/basetrainer.py", line 289, in _train_epoch
>        loss = self.loss_obj(predictions, data)
>      File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 1110, in _call_impl
>        return forward_call(*input, **kwargs)
>      File "/usr/local/lib/python3.7/dist-packages/neuralhydrology/training/loss.py", line 102, in forward
>        {key: prediction[f'{key}{freq_suffix}'] for key in self._prediction_keys},
>      File "/usr/local/lib/python3.7/dist-packages/neuralhydrology/training/loss.py", line 102, in <dictcomp>
>        {key: prediction[f'{key}{freq_suffix}'] for key in self._prediction_keys},
>       KeyError: 'y_hat'

I have changed those 4 parameters but had no result. Could you share some ideas on what to do with it?

Thank you in advance. :)

**Comment by kratzert** [comment, marked as answer]
Is your loss set to `cmalloss`?

**Reply by maby200** [reply]
I have changed it and now it's working. Thank you. :)

**Comment by mengziwen2000** [comment]
```
model:cudalstm
head: cmal
# ----> cmal settings <----
n_distributions: 3
n_samples: 7500
negative_sample_handling: clip
mc_dropout: False
# ----> other settings <----
loss: CMALLoss
target_noise_std: 0.2
output_dropout: 0.5
learning_rate: 
  0: 1e-3
```
In [Uncertainty estimation with DL for rainfall-runoff models](https://arxiv.org/pdf/2012.14295.pdf) .I configured CMAL uncertainty estimates based on cfg arguments, but the results were bad. Roughly what went wrong?
![image](https://github.com/neuralhydrology/neuralhydrology/assets/90944691/bb45847c-46f1-48ac-9128-f2ffcdb8ac5e)

**Reply by gauchm** [reply]
I think the problem lies in the combination of clipping and zero-mean-unit-variance normalization. This is a bit of a problematic thing in neuralhydrology which we should probably document better. 
When setting negative_sample_handling to clip (or truncate) this will do the clipping in the _normalized_ space. So if you use zero-mean/unit-variance normalization, this will cut off everything below the mean of the data. That's why you see this lower bound in the red line of the right plot.

Therefore, your choices are either to turn of clipping, or to turn off the centering of the data (set `centering` to None in `custom_normalization` in the config)

**Reply by mengziwen2000** [reply]
NSE is raised to 0.7 and the hyperparameters can be adjusted again. I'll try to turn off the centering of the data. Thanks.
![image](https://github.com/neuralhydrology/neuralhydrology/assets/90944691/d5c17fca-7a67-48fb-b874-2c4cfc7b25e8)

**Reply by gauchm** [reply]
Note: the issue I described [in my response above](https://github.com/neuralhydrology/neuralhydrology/discussions/88#discussioncomment-6351546) is fixed with v.1.8.1. It is now possible to use centering and `negative_sample_handling` together (e.g., zero-mean/unit-variance scaling and clipping/truncation in the same run).
